File: python/socketd/socketd/transport/stream/impl/StreamBase.py
import asyncio
from abc import ABC
from typing import Callable, Optional

from socketd.exception.SocketDExecption import SocketDTimeoutException
from socketd.transport.core.Message import MessageInternal
from socketd.transport.stream.Stream import StreamInternal
from socketd.transport.stream.StreamManger import StreamManger
from socketd.utils.CompletableFuture import CompletableFuture
from socketd.utils.LogConfig import log
from socketd.utils.RunUtils import RunUtils


class StreamBase(StreamInternal, ABC):
    """流接收器基类
    :param timeout: 超时（毫秒）
    """
    def __init__(self, sid: str, demands: int, timeout: float):
        self.__sid:str = sid
        self.__timeout:float = timeout
        self.__demands:int = demands

        self.__doOnError: Callable[[Exception], None] = None
        self.__doOnProgress: Callable[[bool, int, int], None] = None

        self.__insuranceFuture: Optional[asyncio.Future] = None

    def sid(self) -> str:
        return self.__sid

    def is_done(self):
        return True

    def _then_error_do(self, onError: Callable[[Exception], None]):
        self.__doOnError = onError

    def _then_progress_do(self, onProgress: Callable[[bool, int, int], None]):
        self.__doOnProgress = onProgress

    def demands(self) -> int:
        return self.__demands

    def timeout(self) ->float:
        return self.__timeout

    # 保险开始（避免永久没有回调，造成内存不能释放）
    def insurance_start(self, streamManger: StreamManger, streamTimeout: float) -> None:
        if self.__insuranceFuture:
            return

        async def __insuranceFuture():
            await asyncio.sleep(streamTimeout)
            streamManger.remove_stream(self.__sid)
            self.on_error(SocketDTimeoutException(f"The stream response timeout, sid={self.__sid}"))

        self.__insuranceFuture = CompletableFuture(__insuranceFuture())
        asyncio.run_coroutine_threadsafe(self.__insuranceFuture.get(streamTimeout), asyncio.get_running_loop())

    # 保险取消
    def insurance_cancel(self) -> None:
        if self.__insuranceFuture:
            self.__insuranceFuture.cancel()

    def on_reply(self, reply:MessageInternal):
        ...

    def on_error(self, error: Exception):
        if self.__doOnError:
            RunUtils.taskTry(self.__doOnError(error))
        else:
            log.debug(f"The stream error, sid={self.sid()}", error)

    def on_progress(self, is_send, val, max_val):
        if self.__doOnProgress:
            RunUtils.taskTry(self.__doOnProgress(is_send, val, max_val))

File: javascript/socketd/src/socketd/transport/stream/impl/StreamBase.ts
import {IoConsumer, IoTriConsumer} from "../../core/Typealias";
import {MessageInternal} from "../../core/Message";
import {SocketDTimeoutException} from "../../../exception/SocketDException";
import {Stream, StreamInternal} from "../Stream";
import {StreamManger} from "../StreamManger";

/**
 * 流基类
 *
 * @author noear
 * @since 2.0
 */
export abstract class StreamBase<T extends Stream<any>> implements StreamInternal<T> {
    //保险任务
    private _insuranceFuture: any;
    private _sid: string;
    private _demands: number;
    private _timeout: number;
    private _doOnError: IoConsumer<Error>;
    private _doOnProgress: IoTriConsumer<boolean, number, number>;

    constructor(sid: string, demands: number, timeout: number) {
        this._sid = sid;
        this._demands = demands;
        this._timeout = timeout;
    }

    abstract onReply(reply: MessageInternal);

    abstract isDone(): boolean;


    sid(): string {
        return this._sid;
    }

    demands(): number {
        return this._demands;
    }

    timeout(): number {
        return this._timeout;
    }

    /**
     * 保险开始（避免永久没有回调，造成内存不能释放）
     *
     * @param streamManger  流管理器
     * @param streamTimeout 流超时
     */
    insuranceStart(streamManger: StreamManger, streamTimeout: number) {
        if (this._insuranceFuture) {
            return;
        }

        this._insuranceFuture = setTimeout(() => {
            streamManger.removeStream(this.sid());
            this.onError(new SocketDTimeoutException("The stream response timeout, sid=" + this.sid()));
        }, streamTimeout);
    }

    /**
     * 保险取消息
     */
    insuranceCancel() {
        if (this._insuranceFuture) {
            clearTimeout(this._insuranceFuture);
        }
    }

    /**
     * 异常时
     *
     * @param error 异常
     */
    onError(error: any) {
        if (this._doOnError) {
            this._doOnError(error);
        }
    }

    onProgress(isSend: boolean, val: number, max: number) {
        if (this._doOnProgress) {
            this._doOnProgress(isSend, val, max);
        }
    }

    thenError(onError: IoConsumer<Error>): any {
        this._doOnError = onError;
        return this;
    }

    thenProgress(onProgress: IoTriConsumer<boolean, number, number>): any {
        this._doOnProgress = onProgress;
        return this;
    }
}